Add the `ConcatKdf` procedure

PhilippGackstatter opened the pull request:
# Description of change

Adds the concat KDF as a procedure. See #344 for more details.

Also pulls in some parts from the old stronghold tests to implement a test for the newly introduced functionality. Notably, this also implements `ClientVault::read_secret` for test purposes and is thus behind the `#[cfg(test]` feature.

## Links to any relevant issues

fixes #344 

## Type of change

Choose a type of change, and delete any options that are not relevant.

- [ ] Bug fix (a non-breaking change which fixes an issue)
- [x] Enhancement (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Fix

## How the change has been tested

Added a test similar to the existing X25519 + HKDF test but using the concat KDF instead.

A [test vector from RFC 7518: JSON Web Algorithms](https://www.rfc-editor.org/rfc/rfc7518.html#appendix-C) was added in a unit test.

## Change checklist

Add an `x` to the boxes that are relevant to your changes, and delete any items that are not.

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

PhilippGackstatter committed:
Implement concat kdf procedure

PhilippGackstatter committed:
Merge branch 'dev-refactor' into feat/concat-kdf

PhilippGackstatter committed:
Apply fmt

cycraig commented on client/src/procedures/primitives.rs:
```diff
@@ -789,3 +798,115 @@ impl UseSecret for AeadDecrypt {
+            digest.update(&(apv.len() as u32).to_be_bytes());
+            digest.update(apv);
+
+            // Shared Key Length
+            digest.update(&((len * 8) as u32).to_be_bytes());
+
```
I can't find a reference to including the output length as part of the hash in the NIST PDF (it's not part of `OtherInfo` as far as I can see), nor any existing implementations such as Hyperledger Aries: https://github.com/hyperledger/aries-askar/blob/main/askar-crypto/src/kdf/concat.rs

cycraig replied:
The `keydatalen` is included in the optional `SuppPubInfo` parameter of Concat KDF for JWA with ECDH-ES specifically: https://www.rfc-editor.org/rfc/rfc7518.html#section-4.6.2

>    SuppPubInfo
>       This is set to the keydatalen represented as a 32-bit big-endian
>       integer.
> 
>    SuppPrivInfo
>       This is set to the empty octet sequence.

So it should not be hardcoded, and the procedure should be updated to allow the `SuppPubInfo` and `SuppPrivInfo` optional parameters per the Concat KDF specification.

felsweg-iota commented on client/src/procedures/primitives.rs:
```diff
@@ -789,3 +798,115 @@ impl UseSecret for AeadDecrypt {
+
+impl ConcatKdf {
+    /// The Concat KDF as defined in Section 5.8.1 of NIST.800-56A.
+    pub fn concat_kdf<D: Digest>(
```
Please add more documentation, a short explanation how the algorithm works and it's parameters.

felsweg-iota commented on client/src/procedures/primitives.rs:
```diff
@@ -789,3 +798,115 @@ impl UseSecret for AeadDecrypt {
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct ConcatKdf {
+    /// The hash function to use in the kdf.
+    pub hash: Sha2Hash,
```
Sha-2 Family based hashes are susceptible to [length extension attacks ](https://github.com/marcelo140/length-extension). I would advise to take another digest. blake2b would be a good change. 

Or is SHA256 / X required for this algorithm? I couldn't find an explicit mentioning [here](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-56Ar3.pdf)

cycraig replied:
The SHA256 digest is specified by the JWA RFC 7518 for ECDH: https://www.rfc-editor.org/rfc/rfc7518.html#section-4.6.2

> Key derivation is performed using the Concat KDF, as defined in
>   Section 5.8.1 of [[NIST.800-56A](https://www.rfc-editor.org/rfc/rfc7518.html#ref-NIST.800-56A)], where the Digest Method is SHA-256.

cycraig replied:
Correct me if I'm wrong, but my understanding of the length extension attack is that:

- It requires the/an original, valid hash to perform. Here that would be the output of the Concat KDF itself, which is intended to be a shared secret and thus, if compromised, would mean the attacker would be able to decrypt any related messages anyway, even without performing a length extension attack. 
- It still results in a different hash, and thus would not compromise the derived key here.

The above does make the assumption that the output of the Concat KDF is not used as a MAC, but that seems to be the case for all intended purposes.

I agree that SHA-2 is not ideal, but given that it is specified as a requirement in RFC 7518, we still need to use it to be interoperable with existing implementations.

felsweg-iota replied:
agreed. Looks like there is no other way around.

felsweg-iota reviewed (commented):
Looks good so far, please check the inline comments :)

cycraig commented:
> There are no official test vectors for the concat KDF hence why no unit test itself exists.

There is a test vector in the JWA RFC 7518 with explicit parameters and output bytes, I think we should try match that.

https://www.rfc-editor.org/rfc/rfc7518.html#appendix-C

felsweg-iota commented:
> There is a test vector in the JWA RFC 7518 with explicit parameters and output bytes, I think we should try match that.
@PhilippGackstatter are you providing the impl?

PhilippGackstatter committed:
Add test vector for concat kdf

PhilippGackstatter committed:
Document algorithm; add pub/prv info params

PhilippGackstatter committed:
Simplify digest generics

PhilippGackstatter commented:
Thanks for the comments!

- Added documentation on the `ConcatKDF` struct which is exposed to users, and made the actual function that implements it private. Parameters are documented on the struct and no longer exist for the actual function, so there's less of a need for documentation. But let me know if the documentation is good enough, @felsweg-iota.
- Added the test vector @cycraig mentioned, thanks for finding that!
- Exposed `SuppPubInfo` and `SuppPrivInfo` as parameters, so the impl is no longer specific to how it's used in JWA.

felsweg-iota reviewed (approved):
Thanks @PhilippGackstatter for the changes. Looks good to me.
